src/renderer/shared/api/xcm/service/spellXcmService.ts

```diff
@@ -0,0 +1,256 @@
+import { Builder, Native, convertSs58, getSupportedDestinations } from '@paraspell/sdk-pjs';
+import { type ApiPromise } from '@polkadot/api';
+import { type SubmittableExtrinsic } from '@polkadot/api/types';
+import { BN } from '@polkadot/util';
+
+import { type Asset, AssetType, type Chain } from '@/shared/core';
+import { CHAIN_ID_TO_SPELL_NAME_MAP, formatAmount, toAddress } from '@/shared/lib/utils';
+import { type AccountId } from '@/shared/polkadotjs-schemas';
+
+type XcmTransferParams = {
+  fromChain: Chain;
+  toChain: Chain;
+  asset: Asset;
+  amount: string;
+  destinationAddress: string;
+  senderAddress?: string;
+  fromChainApi?: ApiPromise;
+  toChainApi?: ApiPromise;
+};
+
+type XcmFeeParams = {
+  fromChain: Chain;
+  toChain: Chain;
+  asset: Asset;
+  amount: string;
+  destinationAddress: string;
+  senderAddress?: string;
+  fromChainApi?: ApiPromise;
+  toChainApi?: ApiPromise;
+};
+
+type XcmFeeResult = {
+  originFee: BN;
+  destinationFee: BN | null;
+};
+
+type XcmTransferResult = {
+  extrinsic: SubmittableExtrinsic<'promise'>;
+  disconnect: () => Promise<void>;
+};
+
+type BuilderResult = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  builder: any;
+  fromChainName: string;
+  toChainName: string;
+};
+
+function getSpellChainName(chain: Chain): string | null {
+  return CHAIN_ID_TO_SPELL_NAME_MAP[chain.chainId] ?? null;
+}
+
+function convertAddressToChainFormat(accountId: AccountId, targetChain: Chain, targetChainName: string): string {
+  try {
+    const address = toAddress(accountId, { prefix: targetChain.addressPrefix });
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    return convertSs58(address, targetChainName as any);
+  } catch {
+    return accountId.toString();
+  }
+}
+
+function createCurrencyInput(
+  asset: Asset,
+  amount: string,
+): { amount: string; symbol: string | ReturnType<typeof Native> } {
+  const amountBN = new BN(formatAmount(amount, asset.precision));
+  const isNativeAsset = asset.type === AssetType.NATIVE;
+
+  return {
+    amount: amountBN.toString(),
+    symbol: isNativeAsset ? Native(asset.symbol) : asset.symbol,
+  };
+}
+
+function createBuilderConfig(
+  fromChain: Chain,
+  toChain: Chain,
+  fromChainApi?: ApiPromise,
+  toChainApi?: ApiPromise,
+): { abstractDecimals: boolean; apiOverrides?: Record<string, string[] | ApiPromise> } {
+  const fromChainName = getSpellChainName(fromChain);
+  const toChainName = getSpellChainName(toChain);
+
+  if (!fromChainName || !toChainName) {
+    return { abstractDecimals: false };
+  }
+
+  const apiOverrides: Record<string, string[] | ApiPromise> = {};
+
+  if (fromChainApi) {
+    apiOverrides[fromChainName] = fromChainApi;
+  } else {
+    const fromChainWsUrls = fromChain.nodes.map((node) => node.url);
```

Comment:
Why are we using node urls instead of passing `ApiPromise`? With node urls paraspell will still do a full connect cycle internally